File: src/domain/Order.ts
import OrderItem from './OrderItem';
import {OrderStatus} from './OrderStatus';
import Product from "./Product";

class Order {
  private readonly currency: string = 'EUR';
  private readonly items: OrderItem[] = [];
  private status: OrderStatus = OrderStatus.CREATED;
  private id: number;

  public getTotal(): number {
    return this.items.reduce((acc, item) => acc + item.getVATIncludedAmount(), 0)
  }

  public getCurrency(): string {
    return this.currency;
  }

  public getItems(): OrderItem[] {
    return this.items;
  }

  public getTax(): number {
    return this.items.reduce((acc, item) => acc + item.getTax(), 0)
  }

  public getStatus(): OrderStatus {
    return this.status;
  }

  public setStatus(status: OrderStatus): void {
    this.status = status;
  }

  public getId(): number {
    return this.id;
  }

  public setId(id: number): void {
    this.id = id;
  }

  addItem(product: Product, quantity: number) {
    this.items.push(new OrderItem(product, Quantity.of(quantity)));
  }

}

export default Order;

File: src/main/java/it/gabrieletondi/telldontaskkata/domain/Order.java
package it.gabrieletondi.telldontaskkata.domain;

import it.gabrieletondi.telldontaskkata.domain.status.Created;
import it.gabrieletondi.telldontaskkata.domain.status.OrderStatus;
import it.gabrieletondi.telldontaskkata.service.ShipmentService;
import java.util.ArrayList;
import java.util.List;

public class Order {

  private String currency;
  private List<OrderItem> items;
  private OrderStatus status;
  private int id;

  public Order(int id) {
    this.id = id;
    this.status = new Created();
    this.items = new ArrayList<>();
    this.currency = "EUR";
  }

  public Money getTotal() {
    return items.stream().map(OrderItem::getTaxedAmount).reduce(Money::add).orElse(Money.FREE);
  }

  public String getCurrency() {
    return currency;
  }

  public List<OrderItem> getItems() {
    return items;
  }

  public Money getTax() {
    return items.stream().map(OrderItem::getTax).reduce(Money::add).orElse(Money.FREE);
  }

  public OrderStatus getStatus() {
    return status;
  }

  @Deprecated
  public void setStatus(OrderStatus status) {
    this.status = status;
  }

  public int getId() {
    return id;
  }

  public void addItem(final OrderItem orderItem) {
    items.add(orderItem);
  }

  public void approve() {
    status = status.toApproved();
  }

  public void reject() {
    status = status.toRejected();
  }

  public void ship(final ShipmentService shipmentService) {
    status = status.toShipped();
    shipmentService.ship(this);
  }

  public static class ApprovedOrderCannotBeRejectedException extends RuntimeException {}

  public static class RejectedOrderCannotBeApprovedException extends RuntimeException {}

  public static class ShippedOrdersCannotBeChangedException extends RuntimeException {}

  public static class OrderCannotBeShippedException extends RuntimeException {}

  public static class OrderCannotBeShippedTwiceException extends RuntimeException {}
}
